File: queue/nim/Queue.nim
## Author: Oghogho Odemwingie
## This implementation uses a Doubly Linked List as the underlying data structure

include "../../linkedlist/doubly/nim/DoublyLinkedList"

type
  Queue*[T] = ref QueueObj[T] ## This represents the template for the queue
  QueueObj*[T] = object
    linkedlist : DoublyLinkedList[T] ## The underlying data structure that drives the queue

func newQueue*[T](driver : DoublyLinkedList[T] = newDoublyLinkedList[T]()) : Queue[T] {.inline.} =
  ## Returns a new queue of type `T`
  result = Queue[T](linkedlist : driver)

func enqueue*[T](q : Queue[T], e : T) : void {.inline.} =
  ## Add element `e` to back of the queue
  q.linkedlist.addLast(e)

func dequeue*[T](q : Queue[T]) : T {.inline.} =
  ## Removes and returns the element at the front of the queue
  result = q.linkedlist.removeFirst()

func peek*[T](q : Queue[T]) : T {.inline, raises: [ResourceExhaustedError].} =
  ## Returns the element at the front of the queue
  result = q.linkedlist.peek()

File: queue/julia/Queue.jl
"""
  Author: Oghogho Odemwingie
"""

include("../../linkedlist/doubly/julia/DoublyLinkedList.jl"); using .DoublyLinkedList

"""
This represents the template for the queue
"""
struct Queue{T}
  linkedlist :: LinkedList{T}

  """
  Returns a new queue of type `T`
  """
  Queue{T}(;driver::LinkedList{T} = LinkedList{T}()) where T = new(LinkedList{T}())
end


"""
Add element `e` to back of the queue
"""
function enqueue!(q::Queue{T}, e::T) where T
  addLast!(q.linkedlist, e)
end

"""
Removes and returns the element at the front of the queue
"""
function dequeue!(q::Queue{T}) :: T where T
  removeFirst!(q.linkedlist)
end

Base.peek(q::Queue{T}, ::Type{T}) where T = peek(q.linkedlist)
Base.peek(q::Queue{T}) where T = peek(q.linkedlist)
